src/chk/chk_iv.c

```diff
@@ -84,42 +86,41 @@ chk_iv_ent_update(struct ds_iv_entry *entry, struct ds_iv_key *key,
 {
 	struct chk_iv	*dst_iv = entry->iv_value.sg_iovs[0].iov_buf;
 	struct chk_iv	*src_iv = src->sg_iovs[0].iov_buf;
-	int		 rc = 0;
-
-	/*
-	 * When check leader (IV master) tirgger chk_iv_update, it will set ci_to_leader as 0,
-	 * then chk_iv_ent_update() for any IV message from leader will get -DER_IVCB_FORWARD.
-	 */
-	if (!src_iv->ci_to_leader)
-		D_GOTO(out, rc = -DER_IVCB_FORWARD);
-
-	if (dst_iv->ci_gen == 0)
-		goto update;
-
-	if (unlikely(dst_iv->ci_gen != src_iv->ci_gen)) {
-		D_WARN("Receive invalid update IV message: "DF_X64" vs "DF_X64"\n",
-		       dst_iv->ci_gen, src_iv->ci_gen);
-		goto out;
+	int		 rc;
+
+	if (src_iv->ci_rank == dss_self_rank()) {
+		if (src_iv->ci_to_leader) {
+			/*
+			 * XXX: The case of the check engine sending IV message to the check leader
+			 *	on the same rank has already been handled via chk_iv_update().
+			 */
+			D_ASSERT(!chk_is_on_leader(src_iv->ci_gen, -1, false));
+
+			/* Trigger RPC to the leader via returning -DER_IVCB_FORWARD. */
+			rc = -DER_IVCB_FORWARD;
+		} else {
+			/*
+			 * If it is message to engine, then it must be triggered by leader.
+			 * Return zero that will trigger IV_SYNC to other check engines.
+			 */
+			D_ASSERT(chk_is_on_leader(src_iv->ci_gen, -1, false));
+
+			rc = 0;
+		}
+	} else if (src_iv->ci_to_leader) {
+		*dst_iv = *src_iv;
```

Comment:
D_ASSERT(chk_is_on_leader()) ?

Reply:
Yes, it will be adjusted in subsequent patch.